Standardize running of SimpleReport in Docker

nickclyde opened the pull request:
## Related Issue or Background Info

- Resolves #3295
- Starts #3109

## Changes Proposed

- Create a docker-compose.yml in the root of the repo that runs all parts of the app, including an nginx container that serves the frontend at https://localhost.simplereport.gov/app, the backend at https://localhost.simplereport.gov/api, and serves over https using certs generated by [mkcert](https://github.com/FiloSottile/mkcert)
- This will now be the only supported way to run the app locally. Remove the Makefile and various other docker setups from throughout the README and file structure
- Rewrite `start.sh` to handle cert generation with mkcert and handle running and shutting down docker-compose
- Rewrite README, clearing out old info that no longer applies to the new setup. Especially would like input onto any more README changes y'all are interested in
- Change the pre-commit hooks to use [lefthook](https://github.com/evilmartians/lefthook), since it requires simpler configuration

## Additional Information

- This works for me locally, but I had to literally max out the available resources in Docker Desktop to get it to not be painfully slow (particularly the initial Gradle build on bootup). Wondering if there are any ways to make this more performant? (going to add some `.dockerignore` files to see if it helps speed up builds at least)

## Screenshots / Demos

## Checklist for Author and Reviewer

### Infrastructure
- [ ] **Consult the results of the `terraform-plan` job inside the "Terraform Checks" workflow run for this PR. Confirm that there are no unexpected changes!**

### Design
- [ ] Any UI/UX changes have a designer as a reviewer, and changes have been approved
- [ ] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke-tested by both the engineering and design teams

### Content
- [ ] Any content changes (including new error messages) have been approved by content team

### Support
- [ ] Any changes that might generate new support requests have been flagged to the support team
- [ ] Any changes to support infrastructure have been demo'd to support team

### Testing
- [ ] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- [ ] Database changes are submitted as a separate PR
  - [ ] Any new tables that do not contain PII are accompanied by a GRANT SELECT to the no-PHI user
  - [ ] Any changes to tables that have custom no-PHI views are accompanied by changes to those views
        (including re-granting permission to the no-PHI user if need be)
  - [ ] Liquibase rollback has been tested locally using `./gradlew liquibaseRollbackSQL` or `liquibaseRollback`
  - [ ] Each new changeset has a corresponding [tag](https://docs.liquibase.com/change-types/community/tag-database.html)
- [ ] GraphQL schema changes are backward compatible with older version of the front-end

### Security
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed

## Cloud
- [ ] DevOps team has been notified if PR requires ops support
- [ ] If there are changes that cannot be tested locally, this has been deployed to our Azure `test`, `dev`, or `pentest` environment for verification

emmastephenson commented on README.md:
```diff
@@ -1,32 +1,23 @@
-# Simple Report
+![SimpleReport Logo](/.meta/SimpleReportLogo.svg)
```
nice 👍

emmastephenson commented on README.md:
````diff
@@ -35,195 +26,99 @@ https://simplereport.gov/
-      authorization:
-        granted-roles: ADMIN
-```
+## Support admin
````
Nit: I would switch this section about support admins with the organization roles section below

emmastephenson commented on README.md:
```diff
@@ -35,195 +26,99 @@ https://simplereport.gov/
+- `USER` - a site user the has access to everything in their organization but the gear icon
+- `ENTRY_ONLY` - a site user that only has access to the Conduct Test tab
+- `ALL_FACILITIES` - a site user that can access all facilities in their organization
+- `NO_ACCESS` - a member of an organization who has no permissions without possessing other roles
```
Could you add that all members of an org are part of the no-access group? That's how we get user lists for the organization.

emmastephenson commented on README.md:
```diff
@@ -35,195 +26,99 @@ https://simplereport.gov/
+## API testing
 
-Go to `localhost:8080` to see interact with the graphql api. You would need to point the api endpoint to the backend at: `http://localhost:8080/graphql` This gives you a preview to query/mutate the local database.
+You can make requests to the GraphQL API using [Insomnia](https://insomnia.rest/download) or a similar client. You would need to point the api endpoint to the backend at: `https://localhost.simplereport.gov/api/graphql` This gives you a preview to query/mutate the local database.
```
When using Insomnia, you'll also need to pass your bearer_token in the header - this can be found in Application -> Local Storage -> Access Token on a Chrome browser.

emmastephenson commented on backend/run.sh:
```diff
@@ -0,0 +1,14 @@
+
+###########
+# This script is used for local development only.
+# It will start a continuous build separately from booRun.
```
nit: bootRun?

nickclyde replied:
😅

emmastephenson reviewed (approved):
Overall LGTM, thanks for digging into this one! Just some minor comments on the readme, but the code changes look great.

emmastephenson reviewed (dismissed):
LGTM, thanks for making those README changes!

fzhao99 commented on README.md:
````diff
@@ -35,195 +26,101 @@ https://simplereport.gov/
-```
-touch backend/src/main/resources/application-local.yaml
-```
+How to update resources limits on your [Mac](https://docs.docker.com/desktop/mac/#resources), and [Windows](https://docs.docker.com/desktop/windows/#resources) machines.
````
**Adding note based on pairing session**

On Mac, I needed to give Docker (and my terminal/VS Code for some reason?) access to my disk so it could perform the relevant filesystem operations. Nick mentioned this is might be because my code is housed within my Desktop folder, so leaving this for any future searchers.

Went to System Preferences > Security and Privacy > Privacy and scrolled down to Full Disk Access, granting the relevant apps access to my disk

<img width="660" alt="image" src="https://user-images.githubusercontent.com/29645040/154575811-addd130a-b456-49ba-be3d-446ff8427151.png">

zdeveloper commented on lefthook.yml:
```diff
@@ -0,0 +1,19 @@
+    prettier:
+      root: "frontend/"
+      glob: "*.{js,ts,jsx,tsx}"
+      run: docker run --rm -v ${PWD}:/app/frontend prime-simplereport_frontend sh -c "npx prettier --config /app/frontend/package.json --write {staged_files}" && git add {staged_files}
```
I am still not sure whats the advantage of running linting in docker. you still have to have `yarn`, `npm` and `lefthook` installed on local machine. might as well also install `prettier` `eslint` `stylelint` and run it directly on the machine vs having to rely on docker and installing it everytime via `npx`

nickclyde replied:
Yeah, you make a good point. I think it's a good idea to install the frontend linters locally since we already require node to be installed to run the pre-commit hooks. What do you think about the backend though (spotless)? I haven't required in my instructions to install java locally, but that would be needed to run a grade task locally. Would it make sense to still run that one in docker?

zdeveloper replied:
I think that should also be kept outside docker and on the machine, i don't think its realistic to do app development on the backend without having a jdk installed locally since most IDE's will require it. Intellij requires jdk, it self installs it though. VS code will require it if you install the java extension, its a manual install though.

nickclyde replied:
I guess this sort of just calls the whole approach into question for me. Like, why bother running with docker if I _also_ have to install node and java on my local? At that point, it's just a few commands to have the app running on local. Maybe it would make more sense for the compose setup to just contain the database and nginx, and let the app itself run outside docker?

zdeveloper replied:
I honestly think the docker approach to running it locally is valid, we need to have an easy official way to run the app locally.  An automated way to run both backend & frontend and supporting services. And the Docker setup you made is great for that! The linting is mostly automated and its not part of the `running` part of the development cycle.

jdorothy replied:
Can we roll all of these into a "test" Dockerfile? This file would have all of the dependencies needed for linting/testing, and it could use a local volume to store the dependencies, and then we would use that container for all of these steps. I personally would prefer to not maintain versions of multiple languages on my machine if I can help it.

jdorothy replied:
I think my concern is more surrounding dependency management. I think having a language runtime installed is required if you're developing in that language, so no problems there, but I'd like my dependency management of that language automated and abstracted away from me, if that makes sense.

zdeveloper replied:
Awesome! lets do the test dockerfile approach

zdeveloper reviewed (commented):
LGTM! still unsure why the linting is done in docker

jdorothy commented on README.md:
```diff
@@ -35,195 +26,101 @@ https://simplereport.gov/
-1. run `db-setup/nuke-db.sh`
-2. restart the spring app `gradle bootRun --args='--spring.profiles.active=dev'`
+1. `yarn db` to open up a psql prompt
+1. `DROP DATABASE simple_report; CREATE DATABASE simple_report`
```
This is going to break things with Liquibase and the `simple_report_no_phi` user, since dropping the DB doesn't drop that user and Liquibase is going to try to recreate it once the `databasechangelog` is missing. The ideal thing to do would be to truncate all the tables instead.

nickclyde replied:
I'm going to remove this since it's simple enough to just do a `docker-compose down --volumes`

jdorothy replied:
Perfect!

jdorothy commented on README.md:
```diff
@@ -35,195 +26,101 @@ https://simplereport.gov/
-1. Bring down the service by running `docker-compose down`
-1. Wipe the db by running `docker system prune && docker images prune && docker volume prune`
-1. Restart the service `docker-compose up --build`
+1. `docker system prune --volumes`
```
`docker volume rm backend_simple-report-dev-db-data` (and `docker volume rm simple-report-tests_simple-report-dev-db-data` for tests). Users may have other volumes that they're not interested in pruning at the moment.

You also can't remove a volume from a running container so I'm not sure why `docker-compse up` and `down` are being removed.

nickclyde replied:
Changed to `docker-compose down --volumes`

jdorothy replied:
Perfect!

jdorothy commented on backend/Dockerfile:
```diff
@@ -1,23 +1,16 @@
-FROM gradle:6.7.1-jdk11-hotspot
+FROM gradle:6.9.2-jdk11
```
Is there an issue moving away from the HotSpot version?

alismx replied:
It might be nice to have this and Dockerfile.prod stay in sync as far as which underlying image we pull from.

nickclyde replied:
AFAICT, hotspot is actually the default JVM, so adding `-hotspot` is redundant.

jdorothy replied:
Learn something every day - just checked the list of tags and the hashes are definitely the same.

alismx commented on frontend/.gitignore:
```diff
@@ -15,4 +15,5 @@ build-storybook.log
 
-wiremock-jre8-standalone-2.29.1.jar
\ No newline at end of file
+wiremock-jre8-standalone-2.29.1.jar
```
Would it be alright if we changed this to be more generic in case our versions change `wiremock*` and add it to the dockerignore? It's a sizeable file and will increase the time to setup our build context.

nickclyde commented:
Just pushed a few more changes, which you [can see here](https://github.com/CDCgov/prime-simplereport/pull/3418/files/0088e7d1e6b66affaf36395e78236249d2c8b93b..00868fed33813dad5b609038347556d7b920159b). I addressed the comments above, including adding node modules into the frontend image. I also decided to point our compose file at `ghcr.io` so that we can utilize GitHub's container registry. I added some GH workflows that rebuild our images automatically when dependency files change. This way, instead of rebuilding docker images locally, we can just run `docker-compose pull` to pull down the latest docker images. I added a `docker-compose pull` to the `start.sh` script as well, so it's always up to date. Finally, I updated the pre-commit hooks in lefthook to use the already installed linter binaries!

jdorothy commented on .github/workflows/buildBackendImage.yml:
```diff
@@ -0,0 +1,46 @@
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Extract metadata (tags, labels) for Docker
```
I'm sure I'm missing something obvious but how does the action know that it's building the backend? I don't see a working directory anywhere.

jdorothy replied:
I was in fact missing something obvious: the path comes from the build `context`. In each of these actions, the `context` is `.`, meaning that all 3 actions are trying to build the same `Dockerfile`.

nickclyde replied:
Ooof, you are correct. Good catch. I'll fix that!

jdorothy commented on docker-compose.yml:
```diff
@@ -20,6 +20,7 @@ services:
   backend:
     build:
       context: backend
+    image: ghcr.io/cdcgov/prime-simplereport/backend:latest
```
This is very cool!

emmastephenson reviewed (approved):
LGTM! :shipit:
